### PR TITLE
Support AppendDataExecV1 for Delta Lake tables [databricks]

### LINKED ISSUE #7278: [FEA] Support AppendDataExecV1 for Delta Lake
Delta Lake writes can occur in many forms in a query plan. One form is an AppendDataExecV1. The RAPIDS Accelerator should support Delta Lake writes that use this form in the query plan.

This form appears when using the SQL API to insert into a table e.g.:
```
spark.range(10).createOrReplaceTempView("r")
spark.sql("CREATE TABLE d (id BIGINT) USING DELTA")
spark.sql("INSERT INTO d SELECT * FROM r")
```

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #7278.  Depends on #9443 and #9476.

Implements AppendDataExecV1 for Delta Lake.  The tricky part for this support is that the V1Write instance in the CPU AppendDataExecV1 instance is an anonymous class.  We need to pull out configuration settings in that instance in order to properly translate this to a GPU operation, so reflection was used to extract it.  The mangling of field names by the Scala compiler made this even more fun.

### CODE REVIEW

**Comment by jlowe** [comment]
Marking this draft until the dependency PRs are merged.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala [hunk 1]
```diff
@@ -172,6 +182,116 @@ object DatabricksDeltaProvider extends DeltaProviderImplBase {
...
+      case _ => throw new IllegalStateException("Missing Delta write config from tagging pass")
+    }
+    val deltaLog = writeConfig.deltaLog
+    val gpuWrite = new V1Write {
+      override def toInsertableRelation(): InsertableRelation = {
+        new InsertableRelation {
+          override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+            val session = data.sparkSession
+
+            // TODO: Get the config from WriteIntoDelta's txn.
```
Comment (revans2) [inline delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala:273]: Is this TODO still needed?

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala [hunk 2]
```diff
@@ -172,6 +182,116 @@ object DatabricksDeltaProvider extends DeltaProviderImplBase {
...
+              deltaLog,
+              if (writeConfig.forceOverwrite) SaveMode.Overwrite else SaveMode.Append,
+              new DeltaOptions(writeConfig.options.toMap, session.sessionState.conf),
+              Nil,
+              DeltaLogShim.getMetadata(deltaLog).configuration,
+              data)
+            val gpuWrite = GpuWriteIntoDelta(new GpuDeltaLog(deltaLog, meta.conf), cpuWrite)
+            gpuWrite.run(session)
+
+            // TODO: Push this to Apache Spark
```
Comment (revans2) [inline delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala:284]: Should we file an issue for this?

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala (hunk 1 above)
Comment (jlowe) [reply delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala:273]: This is copied from the original Delta source at https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala#L394

File: delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala (hunk 2 above)
Comment (jlowe) [reply delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala:284]: This is from the original Delta source at https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala#L404-L406

Comment (revans2) [reply delta-lake/common/src/main/databricks/scala/com/nvidia/spark/rapids/delta/DatabricksDeltaProvider.scala:284]: :(

**Review by revans2 — APPROVED** [review]

**Pull request merged**